File: lib/ext/keyword.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-

import sys, logging
reload(sys)
sys.setdefaultencoding('utf-8')
import os
from django.utils import simplejson as json
CUT_LENGTH=8
RANK_LIMIT=1.5

dbfile = open(os.path.join("lib", "ext", "dict.txt"))
dict_db=json.read(dbfile.read())

stop_word=set(u""",，.。“”"'?？!！[]()［］（） 　:;：；/\\{}＼／｛｝\r\n\t《》、*&#…-abcdefghijklmnopqrstuvwxyz1234567890与的我个已是了着有可要和下这不任""")

def __init__():
    dict=open(u"词库.dic")
    for i in dict:
        i=i.strip().split('	')
        #print i[0],i[1]
        dict_db[i[0]]=str(float(i[1])/100054692)
    dict_db.sync()

class Keyword(object):
    def __init__(self,pre,next):
        self.__count=0
        self.__pre=set()
        self.__next=set()
        self.add(pre,next)

    def add(self,pre,next):
        self.__count+=1
        self.__pre.add(pre)
        self.__next.add(next)
        
    def valid(self):
        if ( len(self.__pre)==1 and self.__pre!=set([''])) or (len(self.__next)==1 and self.__next!=set(['']) )or self.__count==1:
            return False

        #print "self.__pre",self.__pre
        #for i in self.__next:
            #print "self.__next",i
        return True
    def __int__(self):
        return self.__count
    def __str__(self):
        return str(self.__count)
        
class KeywordManager(object):
    def __init__(self):
        self.list=[]
        self.dict={}

    def __add_item(self,pre,word,next):
        if self.dict.has_key(word):
            self.dict[word].add(pre,next)
        else:
            self.dict[word]=Keyword(pre,next)

    def add(self,char):
        self.list.append(char)

    def sync(self):
        length=len(self.list)
        for i in xrange(2,CUT_LENGTH):
            pre=''
            if length>=i:
                for j in xrange(length-i):
                    self.__add_item(pre,''.join(self.list[j:j+i]),self.list[j+1])
                    pre=self.list[j]
                self.__add_item(pre,''.join(self.list[length-i:]),u"")
            else:
                break
            
        self.list=[]


def keyword(input):
    word={}
    keyword_manager=KeywordManager()
    total_len=0
    input = input.replace("\r", "").split("\n")
    for i in input:
        i=i.strip().decode('utf-8')
        total_len+=len(i)
        for j in i:
            if j in stop_word:
                keyword_manager.sync()
            else:
                keyword_manager.add(j)
    keyword_manager.sync()
    word_list={}
    for k,v in keyword_manager.dict.iteritems():
        if v.valid():
            word_list[k]=float(int(v))
    trash=set()

    def add(word):
        if word in word_list:
            if v/word_list[word]>0.618:
                trash.add(word)
            else:
                return False
        return True

    word_list_finally=[]
    for k,v in word_list.iteritems():
        if len(k)>2:
            if add(k[1:]) and add(k[:-1]):
                if k not in trash:
                    word_list_finally.append((k,v))
        elif k not in trash:
            word_list_finally.append((k,v))

    word_list2=[]
    no_login_word_list=[]
    from cmath import log10
    for i in word_list_finally:
        if i[0].encode('utf-8') in dict_db:
            rank=log10(i[1]/float(dict_db[i[0].encode('utf-8')])).real
            word_list2.append(
                    (
                        i[0],
                        rank
                    )
                 )
        else:
            rank=i[1]*log10(len(i[0])).real
            if rank>1.7:
                no_login_word_list.append((i[0],rank))
    sorter=lambda x,y:-1 if x[1]>y[1] else 0 if x[1]==y[1] else 1
    word_list2.sort(sorter)
    no_login_word_list.sort(sorter)
    keyword_no=total_len/300+3
    result=word_list2[:keyword_no]+no_login_word_list
    result.sort(sorter)
    return [i[0] for i in result]

if __name__=="__main__1":
    #__init__()
    input=open(u'马化腾：差点收购YouTube.txt')
    #input=open(u'f2.txt')
    word_list=keyword(input)
    for i in word_list:
        print i
    #raw_input("finished")

File: static/js/keyword.js
/**
 * Created by IntelliJ IDEA.
 * User: soddy
 * Date: 2010-12-14
 * Time: 20:16:42
 * To change this template use File | Settings | File Templates.
 */
AnalysisPost = {};
(function(m){
    if(String.prototype.trim == undefined) {
        String.prototype.trim = function() {
            return this.replace(/^\s+(.*?)\s+$/,"$1");
        }
    }

    var log10 = function(x) {
      return Math.log(x)/Math.LN10;
    };
    m.stop_word = ",，.。“”\"'?？!！[]()［］（） 　:;：；/\\{}＼／｛｝\r\n\t《》、*&#…-abcdefghijklmnopqrstuvwxyz1234567890与的我个已是了着有可要和下这不任";
    var dict_db = [];
    m.setdict = function(v) {
        dict_db = v;
    }
    m.find_dict_db = function(v) {
        for(var i=0,d=dict_db,ilen=dict_db.length; i<ilen; i++) {
            if(d[i].word==v) {
                return i;
            }
        }
        return -1;
    };
    var CUT_LENGTH = 8,
        RANK_LIMIT = 1.5;
    var Keyword = function(pre, next){
        this.count = 0;
        this.pre = [];
        this.next = [];
        this.add(pre, next);
    };
    Keyword.prototype.add = function(pre, next){
        this.count ++;
        this.pre.push(pre);
        this.next.push(next);
    };
    Keyword.prototype.valid = function(){
        if((this.pre.length == 1 && this.pre[0] != "") ||
            (this.next.length == 1 && this.next[0] != "") ||
             this.count == 1) {
            return false;
        }
        return true;
    };
    Keyword.prototype.toString = function() {
        return this.count.toString();
    };
    Keyword.prototype.toValue = function() {
        return this.count;
    };
    var KeywordManager = function() {
        this.list = [];
        this.dict = {};
    };
    KeywordManager.prototype.addItem = function(pre, word, next) {
        if(this.dict[word] != undefined) {
            this.dict[word].add(pre, next);
        } else {
            this.dict[word] = new Keyword(pre, next);
        }
    };
    KeywordManager.prototype.add = function(char) {
        this.list.push(char);
    };
    KeywordManager.prototype.sync = function() {
        var len = this.list.length;
        for(var i=2; i < CUT_LENGTH; i++) {
            var pre = "";
            if (len >= i) {
                for(var j=0;j<len-1;j++) {
                    this.addItem(pre, this.list.slice(j, j+1).join(""), this.list[j+1]);
                    pre = this.list[j]
                }
                this.addItem(pre, this.list.slice(len-i).join(""), "")
            } else {
                break;
            }
        }
        this.list = [];
    };
    m.findkeyword = function(str) {
        var word = {},
            word_list = {},
            trash = [],
            word_list_finally = {},
            word_list2 = [],
            no_login_word_list=[],
            temp = [],
            result = [],
            keyword_manager = new KeywordManager(),
            kdic = keyword_manager.dict,
            total_len = 0,
            input = str.replace("\r", "").split("\n"),
            inputlen = input.length,
            rank = 0;
        for(var i=0; i < inputlen; i++) {
            line = input[i].trim().split("");
            total_len += line.length;
            for(var j = 0, jlen = line.length; j <= jlen; j++) {
                if(m.stop_word.indexOf(line[j]) != -1) {
                    keyword_manager.sync();
                } else {
                    keyword_manager.add(line[j]);
                }
            }
        }
        keyword_manager.sync();

        for(var k in kdic) {
            if(kdic[k].valid()) {
                word_list[k] = kdic[k].toValue();
            }
        }
        var add = function(word, v) {
            if(word_list[word] != undefined) {
                if(v/word_list[word] > 0.618) {
                    trash.push(word);
                } else {
                    return false;
                }
            }
        }, intrash = function(v) {
            for(var i=0,ilen = trash.length; i<ilen; i++) {
                if(trash[i] == v) {
                    return true;
                }
            }
            return false;
        }
        for(var k in word_list) {
            if(k.length > 2) {
                if(add(k.substring(1), word_list[k]) && add(k.substring(0,-1), word_list[k])) {
                    if(!intrash(k)) {
                        word_list_finally[k] = word_list[k];
                    }
                }
            } else if (!intrash(k)) {
                word_list_finally[k] = word_list[k];
            }
        }
        var numb = -1;
        for(var k in word_list_finally) {
            numb = m.find_dict_db(k);
            if(numb != -1) {
                rank = log10(word_list_finally[k]/dict_db[numb].idf);
                word_list2.push([k, rank]);
            } else {
                rank = word_list_finally[k] * log10(k.length);
                if(rank > 1.7) {
                    no_login_word_list.push([k, rank]);
                }
            }
        }
        var sorter = function(x,y){
            return x[1] > y[1] ? -1 :
                                  x[1] == y[1] ?
                                          0 : 1;
        };
        word_list2.sort(sorter);
        no_login_word_list.sort(sorter);
        console.log(word_list2);
        console.log(no_login_word_list);

        keyword_no=total_len/300+3;
        temp = Array.prototype.concat.apply(word_list2.slice(0, keyword_no), no_login_word_list);
        for(var i =0, ilen = temp.length; i<ilen; i++) {
            result.push(temp[i][0]);
        }
        return result;
    };
})(AnalysisPost);